File: ruby/prime-factors/prime_factors.rb
module PrimeFactors

  require 'prime'

  def self.for(n)
    primes = Prime.instance.to_enum
    factors = []
    while n > 1
      divisor = primes.next
      while n > 1 && (n % divisor).zero?
        factors << divisor
        n /= divisor
      end
    end
    factors
  end

end

File: elixir/prime-factors/prime_factors.exs
defmodule PrimeFactors do
  @doc """
  Compute the prime factors for 'number'.

  The prime factors are prime numbers that when multiplied give the desired
  number.

  The prime factors of 'number' will be ordered lowest to highest.
  """
  @spec factors_for(pos_integer) :: [pos_integer]
  def factors_for(number) do
    factorize(number, 2, [])
  end

  defp factorize(number, factor, prime_factors) when number < factor do
    prime_factors
  end

  defp factorize(number, factor, prime_factors) when rem(number, factor) == 0 do
    [factor|factorize(div(number, factor), factor, prime_factors)]
  end

  defp factorize(number, factor, prime_factors) do
    factorize(number, factor + 1, prime_factors)
  end

end
